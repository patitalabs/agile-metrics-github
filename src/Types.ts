export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  serviceName: string;
  config: any;
}
