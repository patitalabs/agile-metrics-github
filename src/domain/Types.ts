import {
  ScmCollectorConfig,
  ScmMetricItem,
} from './source-control-management/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface ScmCollectorService {
  fetch(scmCollectorConfig: ScmCollectorConfig): Promise<ScmMetricItem[]>;
}
