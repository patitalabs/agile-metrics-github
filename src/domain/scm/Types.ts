export interface ScmMetricItem {
  sha: string;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  repositoryName: string;
  pullRequest?: {
    prId: number;
    numberOfDaysOpen: number;
    numberOfComments: number;
  };
}

export class ScmCollectorConfig {
  teamName: string;
  repositoryName: string;
  orgName: string;
  since: string;
  until?: string;

  constructor({ repositoryName, orgName, since, until = undefined, teamName }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
