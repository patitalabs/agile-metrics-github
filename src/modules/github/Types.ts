export class GithubCommit {
  sha: string;
  createdAt: Date;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  message: string;
  pullRequest?: PullRequestStats;
}

export class PullRequestStats {
  prId: number;
  numberOfDaysOpen: number;
  numberOfComments: number;
}

export interface GithubConfig {
  repositoryName: string;
  orgName: string;
  since: string;
  until?: string;
}

export interface GithubRepository {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubService {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubClient {
  commits(githubConfig: GithubConfig): Promise<string[]>;

  getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<any>;

  pullRequestForCommit(sha: string): Promise<any>;

  pullRequestComments({
    owner,
    repo,

    number,
  }): Promise<any>;
}
