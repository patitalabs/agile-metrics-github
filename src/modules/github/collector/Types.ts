export interface GithubMetricItem {
  sha: string;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  repositoryName: string;
  pullRequest?: {
    prId: number;
    numberOfDaysOpen: number;
    numberOfComments: number;
  };
}

export class GithubCollectorConfig {
  teamName: string;
  repositoryName: string;
  orgName: string;
  since: string;
  until?: string;

  constructor({ repositoryName, orgName, since, until = undefined, teamName }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
    this.until = until;
    this.teamName = teamName;
  }
}
