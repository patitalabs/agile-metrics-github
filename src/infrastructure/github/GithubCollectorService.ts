import { GithubMetricConverter } from './GithubMetricConverter';
import { ScmCollectorConfig, ScmMetricItem } from '../../domain/source-control-management/Types';
import { GithubService } from './Types';
import { ScmCollectorService } from '../../domain/Types';

export class GithubCollectorService implements ScmCollectorService {
  constructor(private readonly githubService: GithubService) {}

  public async fetch(
    scmCollectorConfig: ScmCollectorConfig
  ): Promise<ScmMetricItem[]> {
    const commits = await this.githubService.commits(scmCollectorConfig);
    return GithubMetricConverter.toMetricItem(commits, scmCollectorConfig);
  }
}
