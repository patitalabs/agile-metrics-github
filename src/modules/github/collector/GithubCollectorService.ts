import { GithubMetricConverter } from './GithubMetricConverter';
import { GithubCollectorConfig, GithubMetricItem } from './Types';
import { GithubService } from '../Types';

export class GithubCollectorService {
  constructor(private readonly githubService: GithubService) {}

  public async fetch(
    githubCollectorConfig: GithubCollectorConfig
  ): Promise<GithubMetricItem[]> {
    const commits = await this.githubService.commits(githubCollectorConfig);
    return GithubMetricConverter.toMetricItem(commits, githubCollectorConfig);
  }
}
