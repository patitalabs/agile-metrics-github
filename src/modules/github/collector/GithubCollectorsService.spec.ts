import { GithubCollectorConfig } from './Types';
import { GithubCommit, GithubConfig, GithubService } from '../Types';
import { GithubCollectorService } from './GithubCollectorService';

function testCommit(): GithubCommit {
  return {
    sha: 'someSha',
    createdAt: new Date('2018-12-06'),
    linesAdded: 30,
    linesRemoved: 6,
    author: 'someAuthor',
    message: 'someMessage',
  };
}

describe('GithubCollectorsService', () => {
  const githubService: GithubService = {
    commits: (githubConfig: GithubConfig): Promise<GithubCommit[]> =>
      Promise.resolve([testCommit()]),
  };

  const githubCollectorsService: GithubCollectorService = new GithubCollectorService(
    githubService
  );

  it('should fetch githubMetrics', async () => {
    const githubCollectorConfig: GithubCollectorConfig = new GithubCollectorConfig(
      {
        teamName: 'someTeamName',
        repositoryName: 'someRepoName',
        orgName: 'someOrgName',
        since: '2018-11-20',
        until: '2020-11-20',
      }
    );

    const data = await githubCollectorsService.fetch(githubCollectorConfig);
    expect(data).toMatchSnapshot();
  });
});
