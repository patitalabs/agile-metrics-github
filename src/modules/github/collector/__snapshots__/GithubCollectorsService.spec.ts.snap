// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubCollectorsService should fetch githubMetrics 1`] = `
Array [
  Object {
    "author": "someAuthor",
    "createdAt": 2018-12-06T00:00:00.000Z,
    "dataType": "SCM",
    "id": "someSha",
    "linesAdded": 30,
    "linesRemoved": 6,
    "message": "someMessage",
    "repositoryName": "someRepoName",
    "sha": "someSha",
    "teamName": "someTeamName",
  },
]
`;
